jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql

```diff
@@ -16,137 +16,275 @@
+    RETURNS ext.mig_num_type[]
+    LANGUAGE plpgsql
+    IMMUTABLE PARALLEL SAFE STRICT
+AS $function$
+DECLARE
+    r ext.mig_num_type;
+    prev_r ext.mig_num_type;
+    h int[];
+    i int;
+    j int;
+BEGIN
+    --parent_num
+    h[-1] = 0;
+    FOR i IN array_lower(rs, 1) .. array_upper(rs, 1)
+        LOOP
+            r = rs[i];
+            h[r.entity_idx_len] = r.num;
+            r.parent_num = h[r.entity_idx_len - 1];
+            rs[i] = r;
+        END LOOP;
+    -- num_cap
+    FOR i IN REVERSE array_upper(rs, 1) .. array_lower(rs, 1)
+        LOOP
+            r = rs[i];
+            IF prev_r is null or r.vo_id != prev_r.vo_id or r.sys_version != prev_r.sys_version THEN
+                --last leaf
+                r.num_cap = r.num;
+                for j in 0 .. r.entity_idx_len LOOP
+                        h[j] = r.num;
+                    END LOOP;
+            ELSEIF  r.entity_idx_len = prev_r.entity_idx_len THEN
+                -- sibling
+                r.num_cap = r.num;
+                h[r.entity_idx_len] = r.num;
+            ELSEIF  r.entity_idx_len < prev_r.entity_idx_len THEN
+                --parent
+                r.num_cap = h[r.entity_idx_len];
+            ELSE -- r.entity_idx_len > prev_r.entity_idx_len
+            --different ancestor
+                r.num_cap = r.num;
+                for j in prev_r.entity_idx_len .. r.entity_idx_len LOOP
+                        h[j] = r.num;
+                    END LOOP;
+            END IF;
+            rs[i] = r;
+            prev_r = r;
+        END LOOP;
+    return rs;
+END;$function$;
+
+--
+-- Select structures from affected table
+--
+CREATE OR REPLACE FUNCTION ext.mig_retrieve_nums_batch(
+    table_name text,
+    id_exp text,
+    ehr_id_exp text,
+    folders_idx_exp text,
+    version_exp text,
+    mig_idx text,
+    batch_size integer
+) RETURNS ext.mig_num_type[]
+    LANGUAGE plpgsql
+    IMMUTABLE STRICT
+AS $function$
+DECLARE
+    rs ext.mig_num_type[];
+BEGIN
+    EXECUTE format('
+        SELECT array_agg(sub.r)
+        FROM(
+            SELECT row(
+                VO_ID,
+                NUM,
+                %5$s,           --EHR_ID
+                ENTITY_IDX_LEN,
+                %3$s,           --SYS_VERSION
+                0,              --parent_num
+                -1              --num_cap
+            ) as r
+            FROM (
+               SELECT %2$s AS id
+               FROM %4$s v
+               WHERE num = 0 and num_cap = -1 and %3$s = 1
+               LIMIT $1
+            ) ids
+            JOIN %4$s d ON ids.id = d.%2$s
+            ORDER BY ids.id, %3$s, d.num
+        ) sub', mig_idx, id_exp, version_exp, table_name, ehr_id_exp, folders_idx_exp)
+        INTO rs
+        USING batch_size;
+    return rs;
+END;$function$;
+
+--
+-- Migration procedure
+--
+CREATE OR REPLACE PROCEDURE ext.mig_num_columns(
+    rel_name regclass,
+    id_exp text,
+    mig_idx text,
```

Comment:
mig_idx is not needed

Reply:
@alexlehn there might have been a misunderstanding:
We do not need the procedure argument, but the indexes still need to be created

Reply:
and they are back again https://github.com/ehrbase/ehrbase/pull/1389/commits/56ee6c1d550d9dc2b3b881f5630e48de5ddd3df1